lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosSobolIndices.cxx

```diff
@@ -294,65 +304,109 @@ Scalar FunctionalChaosSobolIndices::getSobolTotalIndex(const UnsignedInteger var
-          // Now check that all the indices are zero
-          if (*std::max_element(multiIndices.begin(), multiIndices.end()) == 0)
-            covarianceVariables += coefficientI * coefficientI;
-        } // *std::max_element(multiIndices.begin(), multiIndices.end()) > 0
-      } // if coefficientI <> 0
+          // Check that each variable having a nonzero degree is in the group
+          if (multiIndices[variableIndex] > 0 and not variablesGroup.contains(variableIndex)) 
+          {
+            mustInclude = false;
+            break;
+          }
+        }
+      }
+      if (mustInclude) covarianceVariables += coefficientI * coefficientI;
     } // coefficientIndices[i] > 0
   } // Loop over the coefficients
   if (totalVariance > 0.0) return covarianceVariables / totalVariance;
   else return 0.0;
 }
 
 /* Sobol grouped total index accessor */
-Scalar FunctionalChaosSobolIndices::getSobolGroupedTotalIndex(const Indices & variableIndices,
+Scalar FunctionalChaosSobolIndices::getSobolGroupedTotalIndex(const Indices & variablesGroup,
     const UnsignedInteger marginalIndex) const
 {
-  // Compute the complementary indice
   const UnsignedInteger inputDimension = functionalChaosResult_.getDistribution().getDimension();
-  if (!variableIndices.check(inputDimension)) throw InvalidArgumentException(HERE) << "The variable indices of a Sobol indice must be in the range [0, dim-1] and must be different.";
-  const Indices complementaryVariableIndices = variableIndices.complement(inputDimension);
-  // Compute total index from complementary first index
-  const Scalar complementaryFirstIndex = getSobolGroupedIndex(complementaryVariableIndices, marginalIndex);
-  const Scalar groupTotalIndex = 1.0 - complementaryFirstIndex;
-  return groupTotalIndex;
+  if (!variablesGroup.check(inputDimension)) throw InvalidArgumentException(HERE) << "The variable indices of a Sobol' index must be in the range [0, dim-1] and must be different.";
+  const Sample coefficients(functionalChaosResult_.getCoefficients().getMarginal(marginalIndex));
+  const Indices coefficientIndices(functionalChaosResult_.getIndices());
+  const UnsignedInteger size = coefficients.getSize();
+  Scalar covarianceVariables = 0.0;
+  const EnumerateFunction enumerateFunction(functionalChaosResult_.getOrthogonalBasis().getEnumerateFunction());
+  // Sum the contributions of all the coefficients associated to a basis vector involving only the needed variables
+  Scalar totalVariance = 0.0;
+  const UnsignedInteger groupDimension = variablesGroup.getSize();
+  bool mustInclude;
+  for (UnsignedInteger i = 0; i < size; ++i)
+  {
+    if (coefficientIndices[i] > 0)
+    {
+      const Scalar coefficientI = coefficients(i, 0);
+      const Indices multiIndices(enumerateFunction(coefficientIndices[i]));
+      // Compute denominator
+      if (multiIndices.normInf() > 0) totalVariance += coefficientI * coefficientI;
+      // Compute numerator
+      if (multiIndices.normInf() == 0)
```

Comment:
we could try to compute normInf once

Reply:
Ok. Fixed at 3f30713e3d3d5f5de78d890494137f0632a96ab9